torch/_inductor/ir.py

```diff
@@ -4113,6 +4113,123 @@ def codegen(self, wrapper):
         wrapper.writeline(f"_register_tensor_work({output_name}, {output_name}_work)")
 
 
+class MultiOutputNoSizeAssert(MultiOutput):
+    """
+    Extract partial output from a multi-output OP.
+        Works like MultiOutput but doesn't assert size. This must be a property guaranteed by the op emiting this.
+    """
+
+    def codegen(self, wrapper):
+        wrapper.writeline(
+            f"{self.get_name()} = {self.inputs[0].get_name()}{self.index}"
+        )
+
+
+class ForceInPlace(ExternKernel):
+    """
+    Helper OP to encode an in/out argument that tries to make it inplace whenever possible.
+    Wrap the input of your inplace op to enable this behavior.
+
+    TODO: We should test whether wait_tensor can be a victim of reordering and lead to data races.
+    """
+
+    def codegen(self, wrapper):
+        input_name = self.inputs[0].codegen_reference()
+        output_name = self.get_name()
+        if not wrapper.did_reuse(self, self.inputs[0]):
+            wrapper.writeline(f"{output_name}.copy_({input_name}) #no reuse")
```

Comment:
testing my understanding, if it's the `reuse` case, we don't need to do anything as inductor would directly take that variable reference?

I thought `ForceInplace` should actually try to reuse the original variable even if `not wrapper.did_reuse`  returns True?

Reply:
ForceInPlace is a hint to the scheduler (the next PR renamed it to InPlaceHint).

``did_reuse`` here returns whether the input was reused by other, unrelated, ops.

For example:

```python
def kernel_no_reuse(inp):
    buf0 = empty_like(inp) #no reuse
    triton_lalala.run(buf0)

    buf1 = buf0 # reuse it
    dist.all_reduce(buf1)

   buf2 = empty_like(inp)
   triton_lalala.run(buf2)
    return buf1, buf2
```


```python
def kernel_with_reuse(inp):
    buf0 = empty_like(inp) #no reuse
    triton_lalala.run(buf0)

    buf1 = buf0 # reuse it
    dist.all_reduce(buf1)

   buf2 = buf0 #OH NO the scheduler is reusing buf0 instead of allocating a new buffer
   triton_lalala.run(buf2)
    return buf1, buf2
```

The above code will produce the wrong result, so the scheduler plus that codegen check will produce the following instead:


```python
def kernel_with_reuse(inp):
    buf0 = empty_like(inp) #no reuse
    triton_lalala.run(buf0)

    buf1 = empty_like(buf0) # no reuse :( 
    buf1.copy_(buf0)
    dist.all_reduce(buf1)

   buf2 = buf0 #OH NO the scheduler is reusing buf0 instead of allocating a new buffer
   triton_lalala.run(buf2)
    return buf1, buf2
```

This will now produce the right output. The Inductor code is not obvious since allocation happens in the scheduler and is kinda invisible to IR nodes and codegen.